execute/plugin.go

```diff
@@ -114,7 +115,8 @@ func NewPlugin(
 			reportingCfg.OracleID,
 			destChain,
 		),
-		observer: metricsReporter,
+		observer:             metricsReporter,
+		observationOptimizer: exectypes.NewObservationOptimizer(maxObservationLength),
```

Comment:
nit: I'd personally avoid making this service. It seems like there's a single function, so you could make it pure and pass the observation length into the `TruncateObservation` function.

Reply:
I made it like that because we do something similar with reports so maybe in the future extracting an interface with `Truncate` method and any other optimizations we can think about.

Reply:
No strong opinion though

Reply:
Thats a good point that we might want to have multiple implementations. Technically, you already have 2.

Closures are a nice way to sneak configuration into simple handler functions. It's a pattern widely used in a go http middleware libraries:

```
type TruncateExecObservationHandler func(observation) observation

func MakeEncodeEstimateTruncateHandler(maxLength int) TruncateExecObservationHandler {
    return func(observation) observation{
        HashEstimateTruncator(maxLength, observation)
    }
}

func EncodeTruncateHandler(observation) observation {
    // the old function which just has the right signature
}
```

Reply:
Seems like a nice pattern for sure. Let's keep it for next iterations then.

Reply:
One more thing about using a struct is to add other structs and initialize them like the EmptySizesStruct instead of sending it to all the functions or initializing it multiple times.